### ISSUE TITLE
double underscore and double hyphen in image names

### ISSUE THREAD

**Issue by adambrett** [docker/distribution#1056]
Hi All,

I originally raised this issue on the docker repo, but I think this is the more appropriate place.  There was quite a bit of previous discussion there, which for context you can find here: https://github.com/docker/docker/issues/16117

The update to the v2 registry api broke backwards comparability in a minor docker version, as the reduced scope for naming disallowed multiple underscore and hyphens in image names.

I understand the reason for the change, but in order to aid backwards compatibility, I propose allowing double underscore, double hyphen, and double period in image names.

Looking at the work being done in https://github.com/docker/distribution/pull/963, the change would be fairly trivial.

Adding the following test cases to `reference/reference_test.go`:

```go
{
	input:      "test__com",
	repository: "test__com",
},
{
	input:      "test__com--hello",
	repository: "test__com--hello",
},
{
	input:      "test.com/test--com",
	hostname:   "test.com",
	repository: "test.com/test--com",
},
{
	input:      "test--a.com/test..com:sometag",
	hostname:   "test--a.com",
	repository: "test--a.com/test..com",
	tag:        "sometag",
},
{
	input:      "test---a",
	err:   ErrReferenceInvalidFormat,
},
{
	input:      "test___a",
	err:   ErrReferenceInvalidFormat,
},
{
	input:      "test...a",
	err:   ErrReferenceInvalidFormat,
},
{
	input:      "test___a.com/test--com:sometag",
	err:   ErrReferenceInvalidFormat,
},
```

would require the following changes to `reference/regexp.go`:

```diff
--- a/reference/regexp.go
+++ b/reference/regexp.go
@@ -6,7 +6,7 @@ var (
        // nameComponentRegexp restricts registry path component names to
        // start with at least one letter or number, with following parts able to
        // be separated by one or two period, dash or underscore.
-       nameComponentRegexp = regexp.MustCompile(`[a-zA-Z0-9]+(?:[._-][a-z0-9]+)*`)
+       nameComponentRegexp = regexp.MustCompile(`[a-zA-Z0-9]+(?:[._-]{1,2}[a-z0-9]+)*`)
```
I'm open to suggestions from all parties if you need/want more information, or want me to do more investigation into some aspect of this change, as well as how best to integrate it with the work being done in https://github.com/docker/distribution/pull/963.

Thanks

**Comment by dmcgowan** [comment]
I think allowing double hyphen and underscore is a reasonable compromise.  Historically period was no allowed in every part of the name but if there is no technical reason not to allow it we shouldn't add restrictions (https://github.com/docker/docker/blob/a8a31eff10544860d2188dddabdee4d727545796/registry/config.go#L205). Some registries may not allow double hyphen and periods in certain parts of the name, we cannot be expected to enforce that in this package though.

@stevvooe please review and let me know if you have any objection to including this into the reference regexp.

**Comment by dlorenc** [comment]
Any reason to limit this to just two hypens and underscores? We currently have some customers using three.

**Comment by mattmoor** [comment]
Hrm, I just opened: https://github.com/docker/distribution/issues/1062 (thanks for the pointer @dlorenc).

FWIW, it includes a regex proposal, which should meet both our needs.

**Comment by mattmoor** [comment]
I'll ask again here: can we get a fix for this into 1.9?

**Comment by adambrett** [comment]
Was just starting on a PR for this but https://github.com/dmcgowan/distribution/commit/6bd5b8c24e66ba67a54541d72f1ad8d5cc17e677 fixes it, thanks @dmcgowan :)
